src/macros/macrorecorder.cpp

```diff
@@ -0,0 +1,108 @@
+#include "macrorecorder.h"
+
+#include <QtConcurrentRun>
+
+#include "control/controlproxy.h"
+#include "preferences/configobject.h"
+
+// TODO(xerus) handle track eject while recording
+
+MacroRecorder::MacroRecorder()
+        : m_COToggleRecording(ConfigKey(kMacroRecordingKey, "recording_toggle")),
+          m_CORecStatus(ConfigKey(kMacroRecordingKey, "recording_status")),
+          m_activeChannel(nullptr),
+          m_macroRecordingState(MacroState::Disabled),
+          m_pStartRecordingTimer(new QTimer()),
+          m_recordedMacro() {
+    qCDebug(macros) << "MacroRecorder construct";
+
+    connect(&m_COToggleRecording,
+            &ControlPushButton::valueChanged,
+            this,
+            &MacroRecorder::slotToggleRecording);
+    connect(m_pStartRecordingTimer,
+            &QTimer::timeout,
+            this,
+            &MacroRecorder::pollRecordingStart);
+}
+
+void MacroRecorder::notifyCueJump(ChannelHandle& channel, double origin, double target) {
+    qCDebug(macros) << "Jump in channel" << channel.handle();
+    if (checkOrClaimRecording(channel)) {
+        m_recordedMacro.appendJump(origin, target);
+        qCDebug(macros) << "Recorded jump in channel" << channel.handle();
+        setState(MacroState::Armed);
+    }
+}
+
+bool MacroRecorder::checkOrClaimRecording(ChannelHandle& channel) {
+    if (m_activeChannel != nullptr) {
+        return m_activeChannel->handle() == channel.handle() && claimRecording();
+    } else if (claimRecording()) {
+        m_activeChannel = &channel;
+        qCDebug(macros) << "Claimed recording for channel" << channel.handle();
+        return true;
+    }
+    return false;
+}
+
+bool MacroRecorder::claimRecording() {
+    auto armed = MacroState::Armed;
+    return m_macroRecordingState.compare_exchange_weak(armed, MacroState::Recording);
+}
+
+void MacroRecorder::pollRecordingStart() {
+    qCDebug(macros) << "Polling for recording start";
+    if (getState() != MacroState::Disabled) {
+        if (getActiveChannel() == nullptr) {
+            return;
+        } else {
+            ControlProxy(ConfigKey(kMacroRecordingKey, "recording_status")).set(2);
+        }
+    }
+    m_pStartRecordingTimer->stop();
+}
+
+void MacroRecorder::startRecording() {
+    qCDebug(macros) << "MacroRecorder recording armed";
+    m_CORecStatus.set(1);
+    m_recordedMacro.clear();
+    setState(MacroState::Armed);
+    m_pStartRecordingTimer->start(300);
+}
+
+void MacroRecorder::stopRecording() {
+    qCDebug(macros) << "MacroRecorder recording stop";
+    m_CORecStatus.set(0);
+    auto armed = MacroState::Armed;
+    // TODO(xerus) add concurrency test
+    while (!m_macroRecordingState.compare_exchange_weak(armed, MacroState::Disabled))
+        QThread::yieldCurrentThread();
```

Comment:
The max time it can loop is one processing step of the engine. If that takes any measurable length of time, then we have very different problems ^^

Reply:
There is no strict guarantee that the engine sets the value back to armed. Everything that might fail at runtime needs to be handled.

Reply:
What qualifies as a strict guarantee? The codepath that sets it to Recording always ends with setting it back to Armed. Are there memory interferences I am not aware of? Otherwise my only idea would be a crash in the engine thread, which breaks mixxx anyways?

Reply:
No need to discuss this. Two independent threads require proper coordination.

Reply:
What you're trying to do here looks like a "poor mans lock". If we need to reset the recording state to armed to ensure that the macro has been fully written, you should just use SPSCQueue instead.

Reply:
Yeah, basically. I was trying to avoid extra atomics and just have a single unified state.

Reply:
> If we need to reset the recording state to armed to ensure that the macro has been fully written, you should just use SPSCQueue instead.

@xerus2000 ping.

Reply:
Give me some time, there's lots of things to fix ;) I don't forget these, I regularly go through all open comments